go/enclave/enclave_test.go

```diff
@@ -0,0 +1,225 @@
+package enclave
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/crypto/ecies"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/config"
+	"github.com/obscuronet/go-obscuro/go/rpc"
+	"github.com/obscuronet/go-obscuro/go/wallet"
+	"github.com/obscuronet/go-obscuro/integration"
+	"github.com/obscuronet/go-obscuro/integration/datagenerator"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+const _testEnclavePublicKeyHex = "034d3b7e63a8bcd532ee3d1d6ecad9d67fca7821981a044551f0f0cbec74d0bc5e"
+
+var _enclavePubKey *ecies.PublicKey
+
+func init() { //nolint:gochecknoinits
+	// fetch the usable enclave pub key
+	enclPubECDSA, err := crypto.DecompressPubkey(gethcommon.Hex2Bytes(_testEnclavePublicKeyHex))
+	if err != nil {
+		panic(err)
+	}
+
+	_enclavePubKey = ecies.ImportECDSAPublic(enclPubECDSA)
+}
+
+// TestGasEstimation runs the GasEstimation tests
+func TestGasEstimation(t *testing.T) {
+	tests := map[string]func(t *testing.T, w wallet.Wallet, enclave common.Enclave, vk *rpc.ViewingKey){
+		"gasEstimateSuccess":        gasEstimateSuccess,
+		"gasEstimateNoVKRegistered": gasEstimateNoVKRegistered,
+		"gasEstimateNoCallMsgFrom":  gasEstimateNoCallMsgFrom,
+		"gasEstimateInvalidCallMsg": gasEstimateInvalidCallMsg,
+	}
+
+	for name, test := range tests {
+		// create the enclave
+		randomEnclave := createRandomEnclave()
+
+		// create the wallet
+		w := datagenerator.RandomWallet(integration.ObscuroChainID)
+
+		// register the VK with the enclave
+		vk, err := registerWalletViewingKey(t, randomEnclave, w)
+		if err != nil {
+			t.Fatalf("unable to register wallets VK - %s", err)
+		}
+
+		// execute the tests
+		t.Run(name, func(t *testing.T) {
+			test(t, w, randomEnclave, vk)
+		})
+	}
+}
+
+func gasEstimateSuccess(t *testing.T, w wallet.Wallet, enclave common.Enclave, vk *rpc.ViewingKey) {
+	// create the callMsg
+	callMsg := datagenerator.CreateCallMsg()
+	callMsg.From = w.Address()
+	callMsgBytes, err := json.Marshal(callMsg)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// callMsg serialized as a Hex
+	callMsgHex := "0x" + gethcommon.Bytes2Hex(callMsgBytes)
```

Comment:
@otherview I think it would be good to pull reusable bits of code into helper functions to help make the tests more readable. There's lots of hexifying/encrypt/submitting, such that it can hide the purpose of the test.

Reply:
Agree. I don't want to create function to replace a one line action. But as there more endpoints being tested I think helpers will gain a better shape.